src/grain/distribution.js

```diff
@@ -0,0 +1,232 @@
+
+export type DistributionStrategy = ImmediateV1 | LifetimeV1;
+
+export type ImmediateV1 = {|
+  +type: "IMMEDIATE",
+  +version: number,
+  +amount: Grain,
+|};
+
+export type LifetimeV1 = {|
+  +type: "LIFETIME",
+  +version: number,
+  +amount: Grain,
+|};
+
+export type GrainReceipt = {|+address: NodeAddressT, amount: Grain|};
+
+export type DistributionV1 = {|
+  +type: "DISTRIBUTION",
+  +timestampMs: number,
+  +version: number,
+  +strategy: DistributionStrategy,
+  +receipts: $ReadOnlyArray<GrainReceipt>,
+|};
+
+export type CredTimeSlice = {|
+  +intervalEndMs: number,
+  +cred: Map<NodeAddressT, number>,
+|};
+
+export type CredHistory = $ReadOnlyArray<CredTimeSlice>;
+
+/**
+ * Compute a full Distribution given:
+ * - the strategy we're using
+ * - the full cred history for all users
+ * - the lifetime earnings of all users
+ * - the timestamp for the distribution
+ */
+export function distribution(
+  strategy: DistributionStrategy,
+  credHistory: CredHistory,
+  earnings: Map<NodeAddressT, Grain>,
+  timestampMs: number
+): DistributionV1 {
+  const filteredSlices = credHistory.filter(
+    (s) => s.intervalEndMs <= timestampMs
+  );
+  if (!filteredSlices.length) {
+    return {
+      type: "DISTRIBUTION",
+      timestampMs,
+      version: DISTRIBUTION_VERSION_1,
+      strategy,
+      receipts: [],
+    };
+  }
+
+  const receipts: $ReadOnlyArray<GrainReceipt> = (function () {
+    switch (strategy.type) {
+      case "IMMEDIATE":
+        if (strategy.version !== 1) {
+          throw new Error(
+            `Unsupported IMMEDIATE strategy: ${strategy.version}`
+          );
+        }
+        const lastSlice = filteredSlices[filteredSlices.length - 1];
+        return computeImmediateReceipts(strategy.amount, lastSlice.cred);
+      case "LIFETIME":
+        if (strategy.version !== 1) {
+          throw new Error(`Unsupported LIFETIME strategy: ${strategy.version}`);
+        }
+        const totalCred = new Map();
+        for (const {cred} of filteredSlices) {
+          for (const [address, ownCred] of cred.entries()) {
+            const existingCred = totalCred.get(address) || 0;
+            totalCred.set(address, existingCred + ownCred);
+          }
+        }
+        return computeLifetimeReceipts(strategy.amount, totalCred, earnings);
+      default:
+        throw new Error(`Unexpected type ${(strategy.type: empty)}`);
+    }
+  })();
+
+  return {
+    type: "DISTRIBUTION",
+    version: DISTRIBUTION_VERSION_1,
+    strategy,
+    receipts,
+    timestampMs,
+  };
+}
+
+/**
+ * Split a grain amount in proportion to the provided scores
+ */
+function computeImmediateReceipts(
+  distributionAmount: Grain,
+  cred: Map<NodeAddressT, number>
+): $ReadOnlyArray<GrainReceipt> {
+  if (distributionAmount < ZERO) {
+    throw new Error(
+      `invalid distributionAmount: ${String(distributionAmount)}`
+    );
+  }
+
+  const totalCred = sum(cred.values());
+  if (totalCred === 0) {
+    return [];
+  }
```

Comment:
Actually, thinking about this, we might accidentally end up with a noop due to rounding errors.
Perhaps it's better to do a filter at the end, so 0g transactions are removed.

Reply:
I'm not worried about rounding errors producing a 0-attograin distribution, since I doubt we'll ever want single-digit-attograin distributions. I think returning an empty array on a zero distribution makes sense.